clippy_dev/src/lib.rs

```diff
@@ -0,0 +1,157 @@
+extern crate regex;
+#[macro_use]
+extern crate lazy_static;
+extern crate itertools;
+
+use regex::Regex;
+use itertools::Itertools;
+use std::collections::HashMap;
+use std::ffi::OsStr;
+use std::fs;
+use std::io::prelude::*;
+
+lazy_static! {
+    static ref DEC_CLIPPY_LINT_RE: Regex = Regex::new(r#"(?x)
+        declare_clippy_lint!\s*[\{(]\s*
+        pub\s+(?P<name>[A-Z_][A-Z_0-9]*)\s*,\s*
+        (?P<cat>[a-z_]+)\s*,\s*
+        "(?P<desc>(?:[^"\\]+|\\(?s).(?-s))*)"\s*[})]
+    "#).unwrap();
+    static ref DEC_DEPRECATED_LINT_RE: Regex = Regex::new(r#"(?x)
+        declare_deprecated_lint!\s*[{(]\s*
+        pub\s+(?P<name>[A-Z_][A-Z_0-9]*)\s*,\s*
+        "(?P<desc>(?:[^"\\]+|\\(?s).(?-s))*)"\s*[})]
+    "#).unwrap();
+    static ref NL_ESCAPE_RE: Regex = Regex::new(r#"\\\n\s*"#).unwrap();
+    pub static ref DOCS_LINK: String = "https://rust-lang-nursery.github.io/rust-clippy/master/index.html".to_string();
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct Lint {
+    pub name: String,
+    pub group: String,
+    pub desc: String,
+    pub deprecation: Option<String>,
+    pub module: String,
+}
+
+impl Lint {
+    pub fn new(name: &str, group: &str, desc: &str, deprecation: Option<&str>, module: &str) -> Lint {
+        Lint {
+            name: name.to_lowercase(),
+            group: group.to_string(),
+            desc: NL_ESCAPE_RE.replace(&desc.replace("\\\"", "\""), "").to_string(),
+            deprecation: deprecation.map(|d| d.to_string()),
+            module: module.to_string(),
+        }
+    }
+
+    /// Returns all non-deprecated lints
+    pub fn active_lints(lints: &[Lint]) -> impl Iterator<Item=&Lint> {
+        lints.iter().filter(|l| l.deprecation.is_none())
+    }
+
+    /// Returns the lints in a HashMap, grouped by the different lint groups
+    pub fn by_lint_group(lints: &[Lint]) -> HashMap<String, Vec<Lint>> {
+        lints.iter().map(|lint| (lint.group.to_string(), lint.clone())).into_group_map()
+    }
+}
+
+pub fn gather_all() -> impl Iterator<Item=Lint> {
+    lint_files().flat_map(|file| gather_from_file(&file))
+}
+
+fn gather_from_file(dir_entry: &fs::DirEntry) -> impl Iterator<Item=Lint> {
+    let mut file = fs::File::open(dir_entry.path()).unwrap();
+    let mut content = String::new();
+    file.read_to_string(&mut content).unwrap();
+    parse_contents(&content, dir_entry.path().file_stem().unwrap().to_str().unwrap())
+}
+
+fn parse_contents(content: &str, filename: &str) -> impl Iterator<Item=Lint> {
+    let lints = DEC_CLIPPY_LINT_RE
+        .captures_iter(content)
+        .map(|m| Lint::new(&m["name"], &m["cat"], &m["desc"], None, filename));
+    let deprecated = DEC_DEPRECATED_LINT_RE
+        .captures_iter(content)
+        .map(|m| Lint::new( &m["name"], "Deprecated", &m["desc"], Some(&m["desc"]), filename));
+    // Removing the `.collect::<Vec<Lint>>().into_iter()` causes some lifetime issues due to the map
+    lints.chain(deprecated).collect::<Vec<Lint>>().into_iter()
```

Comment:
If I remove the `collect().into_iter()` here, I get `lifetime 'static' required` for `content` and `filename`, which makes sense because they are used in the iterator and consumed outside of this function. I feel like making them static doesn't make sense here and would make the code more complicated to read.